File: Java/quick sort.java
import java.util.Arrays;

public class QuickSort {

    // QuickSort function (recursive)
    public static void quickSort(int[] arr, int low, int high) {
        if (low < high) {
            // Partition the array and get pivot index
            int pi = partition(arr, low, high);

            // Recursively sort elements before and after partition
            quickSort(arr, low, pi - 1);
            quickSort(arr, pi + 1, high);
        }
    }

    // Partition function to place pivot in correct position
    private static int partition(int[] arr, int low, int high) {
        int pivot = arr[high];  // Take last element as pivot
        int i = low - 1;        // Index of smaller element

        // Traverse through all elements
        for (int j = low; j < high; j++) {
            if (arr[j] < pivot) {
                i++;
                // Swap arr[i] and arr[j]
                int temp = arr[i];
                arr[i] = arr[j];
                arr[j] = temp;
            }
        }

        // Swap pivot with element at i+1
        int temp = arr[i + 1];
        arr[i + 1] = arr[high];
        arr[high] = temp;

        return i + 1;  // Return pivot index
    }

    // ✅ Use-case
    public static void main(String[] args) {
        int[] nums = {10, 7, 8, 9, 1, 5};  // Unsorted array

        quickSort(nums, 0, nums.length - 1);  // Perform quicksort

        // Output sorted array
        System.out.println("Sorted: " + Arrays.toString(nums));  // [1, 5, 7, 8, 9, 10]
    }
}

File: C++/quick sort.cpp
#include <iostream>
#include <vector>
using namespace std;

int partition(vector<int>& nums, int low, int high) {
    int pivot = nums[high];  // Choose last element as pivot
    int i = low - 1;         // Index for smaller element

    for (int j = low; j < high; j++) {
        if (nums[j] < pivot) {
            i++;
            swap(nums[i], nums[j]);  // Move smaller element to left
        }
    }
    swap(nums[i + 1], nums[high]);  // Place pivot in correct position
    return i + 1;
}

void quickSort(vector<int>& nums, int low, int high) {
    if (low < high) {
        int pi = partition(nums, low, high);  // Partition index
        quickSort(nums, low, pi - 1);         // Sort left part
        quickSort(nums, pi + 1, high);        // Sort right part
    }
}

// ✅ Use-case
int main() {
    vector<int> nums = {10, 7, 8, 9, 1, 5};
    quickSort(nums, 0, nums.size() - 1);
    for (int x : nums)
        cout << x << " ";  // Output: 1 5 7 8 9 10
}
